osquery/tables/system/linux/portage.cpp

```diff
@@ -0,0 +1,416 @@
+fs::path portageEtcPortageMake = "/etc/portage/make.conf";
+fs::path portageMask = "/etc/portage/package.mask";
+fs::path portageUnMask = "/etc/portage/package.unmask";
+fs::path portageUse = "/etc/portage/package.use";
+fs::path portageWorld = "/var/lib/portage/world";
+
+/* Shared functions and structs */
+
+/**
+ * @brief container to hold data before we store it to QueryData
+ */
+struct portagePackageHolder {
+ public:
+  std::string package;
+  std::string version;
+  std::string keyword;
+  std::string mask;
+  std::string unmask;
+};
+
+/**
+ * @brief split a package string with version into package name and package
+ * version.
+ *
+ * we need to split a package stiring which includes the version it affects into
+ * a pair of package name (first) and package version (second).
+ */
+std::pair<std::string, std::string> portageSplitPackageVersion(
+    std::string pkgStr) {
+  std::pair<std::string, std::string> package = std::make_pair("", "");
+  while (boost::starts_with(pkgStr, ">") || boost::starts_with(pkgStr, "<") ||
+         boost::starts_with(pkgStr, "=")) {
+    if (!(package.second.length() == 0 && boost::starts_with(pkgStr, "="))) {
+      package.second += pkgStr.substr(0, 1);
+    }
+    pkgStr.erase(0, 1);
+  }
+  auto dpos = pkgStr.find('/');
+  auto vpos = pkgStr.find_last_of('-');
+  if (vpos != std::string::npos && vpos > dpos && vpos < pkgStr.length() - 1 &&
+      isdigit(pkgStr[vpos + 1])) {
+    package.second += pkgStr.substr(vpos + 1);
+    pkgStr.erase(vpos, pkgStr.length() - vpos);
+  } else {
+    vpos = pkgStr.find_last_of('-', vpos - 1);
+    if (vpos != std::string::npos && vpos > dpos &&
+        vpos < pkgStr.length() - 1 && isdigit(pkgStr[vpos + 1])) {
+      package.second += pkgStr.substr(vpos + 1);
+      pkgStr.erase(vpos, pkgStr.length() - vpos);
+    }
+  }
+  package.first = pkgStr;
+  return package;
+}
+
+/* Functions to fetch content for database tables */
+QueryData parsePortagePackages(const fs::path& pkgpath,
+                               const std::string& worldContent) {
+  QueryData results;
+
+  fs::path full_path(fs::initial_path());
+  full_path = fs::system_complete(fs::path(pkgpath));
+
+  if (!fs::exists(full_path)) {
+    return results;
+  }
+
+  // store world in a map so we can use find to see if
+  // a package is in the world file
+  std::unordered_map<std::string, bool> world = {};
+  if (worldContent.length() > 0) {
+    for (const auto& i : osquery::split(worldContent, "\n")) {
+      auto line = split(i);
+      if (line.size() == 0 || boost::starts_with(line[0], "#")) {
+        continue;
+      }
+      std::pair<std::string, std::string> package =
+          portageSplitPackageVersion(line[0]);
+      world.emplace(std::make_pair(package.first, true));
+    }
+  }
+
+  // This is a bit slow, would benefit if it had been multithreaded here
+  fs::directory_iterator end_iter;
+  for (fs::directory_iterator dir_itr(full_path); dir_itr != end_iter;
```

Comment:
A note for myself:
I tried to access getGlobb() but not accessible as you are supposed to use resolveFilePattern()